### PR TITLE
Demo app updates for Lite Interpreter: scripts, iOS code, and README

### PR DESCRIPTION
**Pull request by jeffxtang**

### CODE REVIEW

File: ImageSegmentation/README.md [hunk 1]
```diff
@@ -17,11 +17,9 @@ To Test Run the Image Segmentation iOS App, follow the steps below:
 
 ### 1. Prepare the Model
 
-If you don't have the PyTorch environment set up to run the script below to generate the model file, you can download it to the `ios-demo-app/ImageSegmentation` folder using the link [here](https://drive.google.com/file/d/1FHV9tN6-e3EWUgM_K3YvDoRLPBj7NHXO/view?usp=sharing).
+If you don't have the PyTorch environment set up to run the script below to generate the model file, you can download it to the `ios-demo-app/ImageSegmentation` folder using the link [here](https://drive.google.com/file/d/1_guNVutt8eTvO_YhGxkAe1uReBhNaC4f/view?usp=sharing).
```
Comment (chauhang) [inline ImageSegmentation/README.md:20]: Let's move these to a folder under https://download.pytorch.org

File: QuestionAnswering/QuestionAnswering/InferenceModule.mm [hunk 2]
```diff
@@ -57,10 +57,13 @@ - (int)argMax:(NSArray*)array {
             inputs[i] = [tokenIds[i] longValue];
         at::Tensor tensorInputs = torch::from_blob((void*)inputs, {1, MODEL_INPUT_LENGTH}, at::kLong);
         
-        torch::autograd::AutoGradMode guard(false);
```
Comment (chauhang) [inline QuestionAnswering/QuestionAnswering/InferenceModule.mm:60]: Why has this been removed?

File: ObjectDetection/ObjectDetection/Inference/InferenceModule.mm [hunk 3]
```diff
@@ -34,10 +33,13 @@ - (nullable instancetype)initWithFileAtPath:(NSString*)filePath {
 - (NSArray<NSNumber*>*)detectImage:(void*)imageBuffer {
     try {
         at::Tensor tensor = torch::from_blob(imageBuffer, { 1, 3, input_width, input_height }, at::kFloat);
-        torch::autograd::AutoGradMode guard(false);
```
Comment (chauhang) [inline ObjectDetection/ObjectDetection/Inference/InferenceModule.mm:37]: Why has this been removed?

File: ImageSegmentation/README.md [hunk 4]
```diff
@@ -17,11 +17,9 @@ To Test Run the Image Segmentation iOS App, follow the steps below:
...
 
-Be aware that the downloadable model file was created with PyTorch 1.7.0, matching the iOS LibTorch library 1.7.0 specified in the `Podfile`. If you use a different version of PyTorch to create your model by following the instructions below, make sure you specify the same iOS LibTorch version in the `Podfile` to avoid possible errors caused by the version mismatch. Furthermore, if you want to use the latest prototype features in the PyTorch master branch to create the model, follow the steps at [Building PyTorch iOS Libraries from Source](https://pytorch.org/mobile/ios/#build-pytorch-ios-libraries-from-source) on how to use the model in iOS.
```
Comment (chauhang) [inline ImageSegmentation/README.md:22]: Do we need to update all the recipes as well or add new set for lite interpreter specific recipes?

File: ImageSegmentation/ImageSegmentation/TorchModule.mm [hunk 5]
```diff
@@ -40,9 +39,6 @@ - (unsigned char*)segmentImage:(void *)imageBuffer withWidth:(int)width withHeig
 
         at::Tensor tensor = torch::from_blob(imageBuffer, { 1, 3, width, height }, at::kFloat);
 
-        torch::autograd::AutoGradMode guard(false);
```
Comment (chauhang) [inline ImageSegmentation/ImageSegmentation/TorchModule.mm:43]: Why has this been removed?

File: ViT4MNIST/ViT4MNIST/InferenceModule.mm [hunk 6]
```diff
@@ -58,10 +57,13 @@ - (NSString *)recognize:(NSArray*)points {
         
         at::Tensor tensorInputs = torch::from_blob((void*)inputs, {1, 1, MNIST_IMAGE_SIZE, MNIST_IMAGE_SIZE}, at::kFloat);
         
-        torch::autograd::AutoGradMode guard(false);
-        at::AutoNonVariableTypeMode non_var_type_mode(true);
-    
+        c10::InferenceMode guard;
```
Comment (chauhang) [inline ViT4MNIST/ViT4MNIST/InferenceModule.mm:60]: Where is the guard getting used?

File: ViT4MNIST/ViT4MNIST/InferenceModule.mm [hunk 7]
```diff
@@ -58,10 +57,13 @@ - (NSString *)recognize:(NSArray*)points {
         
         at::Tensor tensorInputs = torch::from_blob((void*)inputs, {1, 1, MNIST_IMAGE_SIZE, MNIST_IMAGE_SIZE}, at::kFloat);
         
-        torch::autograd::AutoGradMode guard(false);
```
Comment (chauhang) [inline ViT4MNIST/ViT4MNIST/InferenceModule.mm:61]: Why was this removed?

File: QuestionAnswering/QuestionAnswering/InferenceModule.mm (hunk 2 above)
Comment (xta0) [reply QuestionAnswering/QuestionAnswering/InferenceModule.mm:60]: Once we switched to lite interpreter, we don't need this guard anymore.

**Comment by xta0** [comment]
Thanks @jeffxtang ! This looks great. Have you updated all objective-c wrappers using `c10::InferenceMode guard;`?

File: ImageSegmentation/README.md (hunk 1 above)
Comment (jeffxtang) [reply ImageSegmentation/README.md:20]: How do I access https://download.pytorch.org?

File: ObjectDetection/ObjectDetection/Inference/InferenceModule.mm (hunk 3 above)
Comment (jeffxtang) [reply ObjectDetection/ObjectDetection/Inference/InferenceModule.mm:37]: Same as Tao's answer above. It's replaced by `c10::InferenceMode guard;`

File: ImageSegmentation/README.md (hunk 4 above)
Comment (jeffxtang) [reply ImageSegmentation/README.md:22]: I think so, as all PT Mobile recipes have this:
```
Pre-requisites
PyTorch 1.6.0 or 1.7.0
torchvision 0.6.0 or 0.7.0
```

So do the tutorials https://pytorch.org/tutorials/beginner/deeplabv3_on_ios.html and https://pytorch.org/tutorials/beginner/deeplabv3_on_android.html

File: ImageSegmentation/ImageSegmentation/TorchModule.mm (hunk 5 above)
Comment (jeffxtang) [reply ImageSegmentation/ImageSegmentation/TorchModule.mm:43]: Same as above.

**Comment by jeffxtang** [comment]
> Thanks @jeffxtang ! This looks great. Have you updated all objective-c wrappers to use `c10::InferenceMode guard;`?

Just double checked and found I updated 7 of 8, except D2Go - let me do it now.

**Comment by msaroufim** [comment]
As far as I know downloads.pytorch.org doesn't host model weights but @seemethere can correct me if I'm wrong

The way I've seen models hosted if they come from Facebook is via Torch Hub. You can follow instructions from their README here https://github.com/pytorch/hub

After doing this users can use your model weights with something like

`model = torch.hub.load('pytorch/vision:v0.10.0', 'resnet18', pretrained=True)`

Another alternative is you host all your trained models on an S3 bucket and you can give it a nice name like pytorch-lite-models and people would just wget the S3 url. This is what we do with Torchserve. 

If you'd like to feature the model on the pytorch.org website you can look at uploading it here https://github.com/pytorch/pytorch.github.io

File: ImageSegmentation/README.md (hunk 1 above)
Comment (jeffxtang) [reply ImageSegmentation/README.md]: Done - with replaced S3 links.
